### PR TITLE
fix TxId

### PR DESCRIPTION
**Pull request by zhangyi999**

### CODE REVIEW

File: src/transaction.js [hunk 1]
```diff
@@ -140,18 +143,21 @@ Transaction.prototype.hasWitnesses = function () {
 Transaction.prototype.byteLength = function (stype) {
   let hasWitnesses = this.hasWitnesses()
   let onlyWitnesses = false
+  let hashFull = stype === Transaction.TxSerializeFull
```
Comment (dindinw) [inline src/transaction.js:146]: We don't need the additional `hashFull` variable, use `(hasWitness ? 4 : 0) + // Timestamp` is ok

File: src/transaction.js [hunk 2]
```diff
@@ -115,6 +115,9 @@ Transaction.fromBuffer = function (buffer, __noStrict) {
     }
     tx.locktime = readUInt32()
     tx.exprie = readUInt32()
+  }
+
+  if (tx._stype === Transaction.TxSerializeFull) {
```
Comment (dindinw) [inline src/transaction.js:120]: it's incorrect, the `timestamp` is for ``` tx._stype === Transaction.TxSerializeFull ||
    tx._stype === Transaction.TxSerializeNoWitness) ```, has already done at line 99

File: src/transaction.js [hunk 3]
```diff
@@ -140,18 +143,21 @@ Transaction.prototype.hasWitnesses = function () {
...
   }
   const length =
     4 + // version
-    4 + // Timestamp
     (onlyWitnesses ? 0 : varuint.encodingLength(this.vin.length)) +
     (onlyWitnesses ? 0 : varuint.encodingLength(this.vout.length)) +
     (onlyWitnesses ? 0 : this.vin.reduce(function (sum, input) { return sum + 32 + 4 + 4 }, 0)) + // txid + vout + seq
     (onlyWitnesses ? 0 : this.vout.reduce(function (sum, output) { return sum + 8 + varSliceSize(output.script) }, 0)) + // amount + script
     (onlyWitnesses ? 0 : 4 + 4) + // lock-time + expire
+    (hashFull ? 4 : 0) + // Timestamp
```
Comment (dindinw) [inline src/transaction.js:160]: should be `(hasWitness ? 4 : 0) + // Timestamp`

File: src/transaction.js [hunk 4]
```diff
@@ -212,6 +218,9 @@ Transaction.prototype.toBuffer = function (buffer, initialOffset, stype) {
 
     writeUInt32(this.locktime)
     writeUInt32(this.exprie)
+  }
+
+  if (serializeType === Transaction.TxSerializeFull) {
```
Comment (dindinw) [inline src/transaction.js:223]: same above, incorrect, already done at line 204

File: src/transaction.js (hunk 2 above)
Comment (zhangyi999) [reply src/transaction.js:120]: timestamp Only the hashFull is needed, OnlyWitness is not needed

File: src/transaction.js [hunk 5]
```diff
@@ -220,7 +216,7 @@ Transaction.prototype.toBuffer = function (buffer, initialOffset, stype) {
     writeUInt32(this.exprie)
   }
 
-  if (serializeType === Transaction.TxSerializeFull) {
+  if (serializeType !== Transaction.TxSerializeNoWitness) {
```
Comment (dindinw) [inline src/transaction.js:219]: need not this,  it should handle with other witnesses

**Pull request merged**